### ISSUE TITLE
[REQ] Power "merger" utility for tandem/triplette ride

### ISSUE THREAD

**Issue by Mayeric** [cagnulein/qdomyos-zwift#2882]
**Is your feature request related to a problem? Please describe.**
Doing indoor cycling conjointly with my child or as a couple with my significant other, based on 2 home-trainers, Zwift (or other platform), on a SINGLE (joint) display. But such platform do not propose a "split screen" option + require one subscription for each user. Main solution so far is having 2 completely parallel setup: 2 subscriptions (expensive), 2 computers running the sims, and 2 displays showing different things... a very hardware-intense, costly and not practical solution ! 

**Describe the solution you'd like**
Taking advantage of the "bridging" capability of QZ between the HomeTrainer (HT) world and the Zwift (or other) world, to give it the ability to "merge" the inputs of SEVERAL HT (or a combination of HT / other fitness devices) into a SINGLE "virtual" device that is sent to Z, and represented in Z as a single cyclist with "merged" performances...  Basically a tandem, or a triplette, a quadruplette... 
Could be part of QZ, or maybe a small add-on QZ utility.

Requires :
- being able to connect to at least 2 home-trainers (or other training device)  simultaneously ; ideally maybe up to 3-4 for family use :-))
- merging the performances of the inputs : at least averaging the power and cadence of the 2 HT at every instant... several alternatives may be thought of on the adjust/merge (e.g. for different fitness level)
- Displaying if possible the un-adjusted inputs (Power/cadence) values so that each person can manage his/her own effort
- Send to Z the "merged" Power / cadence, for the simulation to take it into account and display
- Enable the feedback of resistance back from Z to the HTs, if they can handle it. 

Describe alternatives you've considered**
Several "merging" and/or adjustment scheme can be thought of (optional) :
- Simple average between the 2+ HTs (ie. tandem)
- Adjust (e.g. normalize, based on user-inputed FTPmax) the two inputs to a common fitness level - ex- the highest-fit - before merging, so that the users are inputting roughtly the "same" (relative to their FTPmax) level of effort at any given virtual output power. 
- Take the max power between the 2 HTs, as long as the lower power is within 20% (for example) of the max (i.e simulate perfect drafting teaming between 2 cyclists) 
Possible Display (optional):
- unadjusted input, 
-  % of FTPmax
- Visual "contribution" to the team effort : ex. based on individual Power vs. individual FTPmax... could be graphical to identify slacker ! :-) 

**Additional context**

**Comment by cagnulein** [comment]
I created a proof of concept script 

You need to install first

pip install bleak bluez-peripheral

`import asyncio
from bleak import BleakScanner, BleakClient
from bleak.backends.scanner import AdvertisementData
from bleak.backends.device import BLEDevice
from typing import Dict, List
import statistics
from bluez_peripheral.gatt.service import Service
from bluez_peripheral.gatt.characteristic import characteristic
from bluez_peripheral.util import *
from bluez_peripheral.advert import Advertisement
import logging
import datetime

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - [%(name)s] %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)

class DeviceLogger:
    def __init__(self, name: str, address: str):
        self.logger = logging.getLogger(f"{name}_{address}")
    
    def log_data(self, direction: str, data_type: str, data: bytearray):
        hex_data = ' '.join([f"{b:02x}" for b in data])
        self.logger.info(f"{direction} {data_type}: {hex_data}")

class FTMSService(Service):
    def __init__(self):
        super().__init__(FTMS_SERVICE_UUID)
        self._power = 0
        self._cadence = 0
        self.logger = DeviceLogger("VirtualBridge", "00:00:00:00:00:00")

    @characteristic(INDOOR_BIKE_DATA_UUID, notify=True)
    def indoor_bike_data(self, options):
        data = bytearray([0x18])  # Power and Cadence enabled
        data.extend([0x00, 0x00])  # Speed
        data.extend(self._power.to_bytes(2, 'little'))
        data.extend(self._cadence.to_bytes(2, 'little'))
        self.logger.log_data("OUT", "INDOOR_BIKE_DATA", data)
        return data

    @characteristic(FTMS_CONTROL_POINT_UUID, write=True, notify=True)
    def control_point(self, value):
        self.logger.log_data("IN", "CONTROL_POINT", value)
        asyncio.create_task(self._broadcast_control_point(value))
        return value

    async def _broadcast_control_point(self, value):
        if hasattr(self, 'bridge'):
            await self.bridge.broadcast_control_point(value)

    def update_metrics(self, power: int, cadence: int):
        self._power = power
        self._cadence = cadence
        self.indoor_bike_data.changed()

class FTMSBridge:
    def __init__(self):
        self.devices: Dict[str, BleakClient] = {}
        self.device_loggers: Dict[str, DeviceLogger] = {}
        self.power_values: Dict[str, int] = {}
        self.cadence_values: Dict[str, int] = {}
        self.service = FTMSService()
        self.service.bridge = self
        self.advertisement = Advertisement("FTMS Bridge", [FTMS_SERVICE_UUID])

    async def broadcast_control_point(self, value: bytearray):
        for address, client in self.devices.items():
            logger = self.device_loggers[address]
            try:
                logger.log_data("OUT", "CONTROL_POINT", value)
                await client.write_gatt_char(FTMS_CONTROL_POINT_UUID, value)
            except Exception as e:
                logger.logger.error(f"Failed to write control point: {e}")

    async def scan_and_connect(self):
        scanner = BleakScanner()
        scanner.register_detection_callback(self._device_detected)
        await scanner.start()
        return scanner

    async def _device_detected(self, device: BLEDevice, adv: AdvertisementData):
        if (FTMS_SERVICE_UUID.lower() in adv.service_uuids and 
            device.address not in self.devices):
            
            logger = DeviceLogger(f"Device_{len(self.devices)+1}", device.address)
            logger.logger.info(f"Found device: {device.name} ({device.address})")
            
            client = BleakClient(device)
            try:
                await client.connect()
                self.devices[device.address] = client
                self.device_loggers[device.address] = logger
                logger.logger.info("Connected successfully")
                await self._start_notifications(client)
            except Exception as e:
                logger.logger.error(f"Failed to connect: {e}")

    async def _start_notifications(self, client: BleakClient):
        logger = self.device_loggers[client.address]
        try:
            await client.start_notify(
                INDOOR_BIKE_DATA_UUID,
                lambda _, data: self._handle_bike_data(client.address, data)
            )
            logger.logger.info("Started notifications")
        except Exception as e:
            logger.logger.error(f"Failed to start notifications: {e}")

    def _handle_bike_data(self, address: str, data: bytearray):
        logger = self.device_loggers[address]
        logger.log_data("IN", "INDOOR_BIKE_DATA", data)
        
        if len(data) >= 7:
            flags = data[0]
            if flags & 0x18:
                power = int.from_bytes(data[3:5], 'little')
                cadence = int.from_bytes(data[5:7], 'little')
                self.power_values[address] = power
                self.cadence_values[address] = cadence
                logger.logger.info(f"Power: {power}W, Cadence: {cadence}rpm")
                self._update_virtual_device()

    def _update_virtual_device(self):
        if self.power_values and self.cadence_values:
            avg_power = int(statistics.mean(self.power_values.values()))
            avg_cadence = int(statistics.mean(self.cadence_values.values()))
            self.service.logger.logger.info(
                f"Broadcasting averages - Power: {avg_power}W, Cadence: {avg_cadence}rpm"
            )
            self.service.update_metrics(avg_power, avg_cadence)

    async def run(self):
        logger = logging.getLogger("FTMSBridge")
        logger.info("Starting FTMS Bridge...")
        scanner = await self.scan_and_connect()
        try:
            while True:
                await asyncio.sleep(1)
        except KeyboardInterrupt:
            logger.info("Shutting down...")
            await scanner.stop()
            for address, client in self.devices.items():
                logger = self.device_loggers[address]
                logger.logger.info("Disconnecting...")
                await client.disconnect()

if __name__ == "__main__":
    bridge = FTMSBridge()
    asyncio.run(bridge.run())`

**Comment by cagnulein** [comment]
[ftms-bridge.zip](https://github.com/user-attachments/files/18027507/ftms-bridge.zip)

**Comment by Mayeric** [comment]
Thanks a lot ! I will have a try at this, but give me some time as I am not from the "dev" world (1st time github user,, 2nd time Python user...), so I need a bit of set-up and ramping-up...  (I understand this could be tested as a Python script on a Windows-based Python set-up with a Bluetooth dongle).

**Comment by cagnulein** [comment]
Yes you need python and whatever operating system with a Bluetooth device
I didn't test it myself so share a log when you will try it!

Roberto Viola
Software engineer and open source enthusiast
http://robertoviola.cloud



Il giorno sab 7 dic 2024 alle 08:55 Mayeric ***@***.***> ha
scritto:
